blog/02-18-2023-Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain/Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain.md

```diff
@@ -0,0 +1,998 @@
+
+`canSwallow` This modifier checks whether the power value of the attacker is greater than the power value of the owner of the NFT before allowing the `swallowNFT()` function to execute. If the power value of the attacker is less than or equal to the owner's power value, the function execution is aborted.
+
+let's look at the functions
+
+```solidity
+ function mint( string memory name) public payable {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(msg.sender, tokenId);
+        addNFT(tokenId, name);// listing the nft 
+    }
+```
+
+Let's declare a function called `mint()` The Mint function allows a player to mint their own NFTs. The function takes a string as a parameter which is used to specify the name of the NFT. The function first generates a unique `tokenId` for the NFT and increments the `tokenId counter`. It then calls the `_safeMint` function which is inherited from the `ERC721` contract and is used to mint the NFT. Finally, it calls the `addNFT` function which adds the NFT to the game.
+
+```solidity
+ function addNFT(uint256 _tokenId, string memory _name) private{
+        uint _powerValue = 0;
+        nfts[allNFTs] = NFT(
+            _tokenId,
+            payable(msg.sender),
+            _powerValue,
+            _name
+        );
+        allNFTs++;
+        minters[msg.sender] = true;
+    }
+```
+
+The `AddNFT()` function adds the NFT to the game. The function takes a `tokenId` and a name as parameters. It then creates a new NFT struct with the given parameters and adds it to the internal nfts mapping. It also adds the player to the minters mapping and increments the `allNFTs` variable. 
+
+```solidity
+function swallowNFT(uint _index) external hasmint(msg.sender) canSwallow(msg.sender, _index){
+	       require(msg.sender != nfts[_index].owner, "can't swallow your own nft");         
+           _transfer(nfts[_index].owner, msg.sender, nfts[_index].tokenId);
+           playerpowervalue[nfts[_index].owner] -= nfts[_index].powerValue;
+           playerpowervalue[msg.sender] += nfts[_index].powerValue;
+           nfts[_index].owner = payable(msg.sender);
+  }
+```
+
+The `SwallowNFT()` function allows a player to swallow another player’s NFT. This function takes an index as a parameter which is used to specify the NFT to be swallowed. The function is protected by two modifiers. `hasmint` and `canSwallow`. The `hasmint` modifier checks if the address has minted an NFT and the `canSwallow` modifier checks if the power value of the attacker is greater than the power value of the owner. If both the modifiers pass, the function transfers the NFT from the owner to the attacker, decreases the power value of the owner and increases the power value of the attacker. 
```

Comment:
The `SwallowNFT()` function allows a player to swallow another player’s NFT. This function takes an index as a parameter that specifies the NFT to be devoured. The function is protected by two modifiers. `hasmint` and `canSwallow`. The `hasmint` modifier checks if the address has minted an NFT, and the `canSwallow` modifier checks if the power value of the attacker is greater than the power value of the owner. If both modifiers pass, the function transfers the NFT from the owner to the attacker, decreases the owner's power value and increases the attacker's power value.